### PR TITLE
[WIP] feature nftpar 1491 add tests

### PR DESCRIPTION
**Pull request by Degtiv**
DO NOT MERGE!

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -1,6 +1,15 @@
...
 POSTGRES_PORT=5432
+
+#Environment variables for tests
+TESTS_WS_PROVIDER_URL='ws://host.docker.internal:9944'
+TESTS_UNIQUE_WS_ENDPOINT=wss://ws-quartz-dev.unique.network
+TESTS_TYPE_PROVIDER=quartz
+TESTS_GRAPHQL_URL='http://localhost:3031/graphql'
+TESTS_CHAIN_VERSION=918001
+TESTS_CRAWLER_API_REPOSITORY=https://github.com/UniqueNetwork/unique-explorer-api/
+TESTS_CRAWLER_BRANCH=v1.6.3
```
Comment (Maksandre) [inline .env:15]: At least these three variables would be redundant when you will fix your running scripts

File: tests/api/collections.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import { collectionsQuery } from '../queries';
+import axios from 'axios';
+import 'dotenv/config';
+import { handleResponse } from './responseHandlers';
+
+const scanApi = axios.create({
+  baseURL: process.env.GRAPHQL_URL,
```
Comment (Maksandre) [inline tests/api/collections.ts:7]: As I can see you don't have this variable in your configuration.

File: tests/scripts/create-test-collections.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+const { collections } = require('../data/collections');
+const config = require('../utils/unique/config');
+
+// Works only for quartz >= 921080
+const mintCollectionsAndTokens = async () => {
+  // eslint-disable-next-line no-undef
+  if (process.env.UNIQUE_COLLECTION_IDS) {
+    console.log('UNIQUE_COLLECTION_IDS is defined');
+    return;
+  }
```
Comment (Maksandre) [inline tests/scripts/create-test-collections.js:15]: Are you sure you need all this script?

File: tests/data/collections.js [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+const collections = [
+  {
+    schema: {
+      name: 'New Substra Punks',
+      description: 'Like old punks',
+      tokenPrefix: 'PNK',
+      properties: [
+        {
+          key: '_old_offchainSchema',
+          value:
```
Comment (Maksandre) [inline tests/data/collections.js:12]: Are you sure you need this file?

File: tests/scripts/prepare-scan-api.sh [hunk 5]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
```
Comment (Maksandre) [inline tests/scripts/prepare-scan-api.sh:1]: This script should be improved – more info in the comments for package.json

File: tests/scripts/run-chain.sh [hunk 6]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
```
Comment (Maksandre) [inline tests/scripts/run-chain.sh:1]: This script should be improved – more info in the comments for package.json

File: tests/scripts/run-scan-api.sh [hunk 7]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+
+ENV_FILE=${1}
+export $(grep -v '^#' $ENV_FILE | xargs)
```
Comment (Maksandre) [inline tests/scripts/run-scan-api.sh:4]: This script should be improved – more info in the comments for package.json

File: tests/utils/helpers.ts [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+  let str = '';
+  for (let i = 0, strLen = buf.length; i < strLen; i++) {
+    if (buf[i] != 0) {
+      str += String.fromCharCode(buf[i]);
+    } else {
+      break;
+    }
+  }
+  return str;
+}
```
Comment (Maksandre) [inline tests/utils/helpers.ts:21]: You can now use str2vec and vec2str from playgrounds UniqueUtil class

File: tests/utils/retry.ts [hunk 9]
```diff
@@ -0,0 +1,20 @@
+export const retry = async (callback: () => Promise<any>, timeout = 3000) => {
```
Comment (Maksandre) [inline tests/utils/retry.ts:1]: Maybe we need to add some sleep between retries.

File: tests/utils/settings.ts [hunk 10]
```diff
@@ -0,0 +1,26 @@
+export interface Settings {
```
Comment (Maksandre) [inline tests/utils/settings.ts:1]: You don't import this file, so do you actually need it?

**Review by Maksandre — CHANGES_REQUESTED** [review]

File: tests/utils/settings.ts (hunk 10 above)
Comment (Degtiv) [reply tests/utils/settings.ts:1]: done

File: tests/utils/retry.ts (hunk 9 above)
Comment (Degtiv) [reply tests/utils/retry.ts:1]: done

File: tests/utils/helpers.ts (hunk 8 above)
Comment (Degtiv) [reply tests/utils/helpers.ts:21]: done

File: tests/api/collections.ts (hunk 2 above)
Comment (Degtiv) [reply tests/api/collections.ts:7]: done

File: .env (hunk 1 above)
Comment (Degtiv) [reply .env:15]: done

File: .gitignore [hunk 11]
```diff
@@ -17,6 +17,7 @@ lerna-debug.log*
 # Tests
 /coverage
 /.nyc_output
+/tests/application/*
```
[the remark this answers is not in the document]
Comment (Degtiv) [reply .gitignore:20]: done

File: package.json [hunk 12]
```diff
@@ -13,12 +13,16 @@
...
     "start:debug": "nest start --debug --watch",
     "start:prod:web-api": "node dist/apps/web-api/main ",
-    "lint": "eslint \"{src,apps,libs,test,migrations}/**/*.ts\" --fix",
+    "lint": "eslint \"{src,apps,libs,test,tests,migrations}/**/*.ts\" --fix",
     "test": "jest",
     "test:watch": "jest --watch",
     "test:cov": "jest --coverage",
     "test:debug": "node --inspect-brk -r tsconfig-paths/register -r ts-node/register node_modules/.bin/jest --runInBand",
     "test:e2e": "jest --config ./apps/web-api/test/jest-e2e.json",
+    "tests:run-chain": "(cd tests/scripts ; ./run-chain.sh ../../.env)",
```
[the remark this answers is not in the document]
Comment (Degtiv) [reply package.json:22]: done

File: tests/scripts/run-chain.sh (hunk 6 above)
Comment (Degtiv) [reply tests/scripts/run-chain.sh]: done

File: package.json [hunk 13]
```diff
@@ -13,12 +13,16 @@
...
+    "tests:prepare-scan-api": "(cd tests/scripts ; ./prepare-scan-api.sh ../../.env)",
+    "tests:run-scan-api": "(cd tests/scripts ; ./run-scan-api.sh ../../.env)",
+    "tests:run-environment": "npm run tests:run-chain ; npm run tests:run-scan-api",
```
[the remark this answers is not in the document]
Comment (Degtiv) [reply package.json]: done

File: tests/scripts/run-scan-api.sh (hunk 7 above)
Comment (Degtiv) [reply tests/scripts/run-scan-api.sh]: done

File: e2e/utils/unique/config.js [hunk 14]
```diff
@@ -0,0 +1,10 @@
+/* eslint-disable @typescript-eslint/no-var-requires */
```
Comment (Maksandre) [inline e2e/utils/unique/config.js:1]: Do you need utils/unique at all? I can't find any references to all these files

File: e2e/utils/expect.ts [hunk 15]
```diff
@@ -0,0 +1,23 @@
+import { expect } from 'chai';
```
Comment (Maksandre) [inline e2e/utils/expect.ts:1]: Does this code duplicate the one written in retry?

File: package.json [hunk 16]
```diff
@@ -31,6 +34,7 @@
     "@nestjs/graphql": "^9.1.2",
     "@nestjs/platform-express": "^8.2.5",
     "@nestjs/typeorm": "^8.0.3",
+    "@unique-nft/sdk": "^0.3.16",
```
Comment (Maksandre) [inline package.json:37]: Shouldn't this package be added to devDependencies instead?

File: docker-compose-e2e.yml [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+    networks:
+      - scan
+    restart: on-failure
+    ports:
+      - "9933:9933"
+      - "9944:9944"
+      - "9844:9844"
+
+  crawler-migrations:
+    image: registry.uniquenetwork.dev/uniquenetwork/unique-explorer-api:b66f5dc00b4de9446bd8591f5888497a510f9c2a35a01c3dda1d6616-1655272711166
```
Comment (Maksandre) [inline docker-compose-e2e.yml:78]: We need to find a way to avoid using the registry. Please add it to a backlog.

**Review by Maksandre — APPROVED** [review]

File: docker-compose-e2e.yml [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+  blockchain:
+    image: uniquenetwork/quartz-node-private:latest
+    container_name: scan-blockchain
+    networks:
+      - scan
+    restart: on-failure
+    ports:
+      - "9933:9933"
+      - "9944:9944"
+      - "9844:9844"
```
Comment (Maksandre) [inline docker-compose-e2e.yml:75]: Let's comment out these lines until the local chains are fixed?

File: e2e/utils/expect.ts (hunk 15 above)
Comment (Degtiv) [reply e2e/utils/expect.ts:1]: I've delete retry.ts

File: package.json (hunk 16 above)
Comment (Degtiv) [reply package.json:37]: moved

File: e2e/utils/unique/config.js (hunk 14 above)
Comment (Degtiv) [reply e2e/utils/unique/config.js:1]: deleted

File: docker-compose-e2e.yml (hunk 18 above)
Comment (Degtiv) [reply docker-compose-e2e.yml:75]: commented

**Comment by Degtiv** [comment]
I broke this branch using rebase. I've create new clear branch and PR: https://github.com/UniqueNetwork/unique-scan-backend/pull/36

**Pull request closed without merging**